docs/typescript/reference/plugins.md

```diff
@@ -0,0 +1,172 @@
+---
+sidebar_position: 200
+title: DBOS Plugin Architecture
+---
+
+# DBOS Plugin Architecture
+
+:::tip
+Unless you intend to extend the DBOS Transact library, you can ignore this topic.
+:::
+
+DBOS Transact for TypeScript currently provides two explicit extension mechanisms:
+* [Datasources](#datasources) integrate database clients, ORMs, or other resouces with the DBOS lifecycle and transaction functions.
+* [External Event Receivers](#event-receivers) integrate event receivers (pollers, inbound sockets, timers, etc.) with the DBOS lifecycle, and initate workflow functions upon receiving events.
+
+## Datasources
+
+### Method Of Operation
+
+The principle behind DBOS datasources is that DBOS and the datasource work together to execut functions in a checkpointed, transactional context.
+1. App code calls a transaction function (which may or may not be registered in advance).  Generally, this call is made to the datasource instance, as it provides full type checking for the transaction configuration.
+2. DBOS runs its step wrapper first, consulting the system database, and skipping the step if it has already run.
+3. If the step should be run, DBOS calls the datasource to start a transaction.
+4. The datasource starts a transaction and calls the app's function, making the transactional datasource client available to that function via context.
+5. Upon completion of the function, the datasource generally writes its own completion checkpoint before ending the transaction.  Including this checkpoint in the transaction ensures exactly-once processing of the transaction.
+6. Control returns to the DBOS step wrapper, which writes the step result to its system database and emits tracing information.
+7. The transaction's result is returned to the caller.
+
+### Datasource Interfaces
+
+Datasource integration is based on 3 key interfaces:
+ - `DBOSDataSource`: User-facing datasource API.  Part interface and part guideline, this is technically just a suggestion, but following it ensures a consistent developer experience across datasources.
+ - `DataSourceTransactionHandler`: This interface is provided by the datasource to DBOS, so that DBOS can control transaction scope.
+ - `@dbos-inc/dbos-sdk/datasource`: Exports from this submodule are used by datasources to do register themselves and initiate transactions.
+
+#### `DBOSDataSource`
+
+`DBOSDataSource` is the API for datasources to present to the application developer.  It provides a consistent, type-safe way to register and call transaction functions.
+`DBOSDataSource` is templatized on its `Config` type, which describes datasource-specific configuration for a transaction function.  `Config` typically covers transaction isolation, read-write/read-only, and the transaction's name.
+
+ - `runTransaction` should run its callback, generally with the `runTransaction` function [below](#dbos-incdbos-sdkdatasource)
```

Comment:
I don't like how we keep saying "should" and "generally". Maybe something more like: The `DBOSDataSource.runTransaction` method is used to execute arbitrary database code and save the return value in a single transaction. DBOS provides the `runTransaction` helper function described below that can be used when implementing `DBOSDataSource.runTransaction`.